trace-dispatcher/src/Cardano/Logging/ConfigurationParser.hs

```diff
@@ -19,53 +20,108 @@ import qualified Data.ByteString.Char8 as BS
+    , backends :: Maybe [BackendConfig]
+    , maxFrequency :: Maybe Double
+    }
+  deriving (Eq, Ord, Show, Generic)
+
+instance AE.FromJSON ConfigOptionRep where
+  parseJSON (Object obj) = ConfigOptionRep
+                         <$> obj .:? "severity"
+                         <*> obj .:? "detail"
+                         <*> obj .:? "backends"
+                         <*> obj .:? "maxFrequency"
+  parseJSON _ = mempty
+
+instance AE.ToJSON ConfigOptionRep where
+  toJSON ConfigOptionRep{..} = object (conss [])
+    where
+      consMay attr = maybe id ((:) . (attr .=))
+      conss = consMay "severity" severity
+            . consMay "detail" detail
+            . consMay "backends" backends
+            . consMay "maxFrequency" maxFrequency
+
+type OptionsRepresentation = Map.Map Text ConfigOptionRep
+
+instance AE.ToJSON TraceConfig where
+  toJSON tc = toJSON (configToRepresentation tc)
+
+-- | Read a configuration file and returns the internal representation
 readConfiguration :: FilePath -> IO TraceConfig
 readConfiguration fp =
     either throwIO pure . parseRepresentation =<< BS.readFile fp
 
+-- | Read a configuration file and returns the internal representation
+-- Uses values which are not in the file fram the defaultConfig
 readConfigurationWithDefault :: FilePath -> TraceConfig -> IO TraceConfig
 readConfigurationWithDefault fp defaultConf = do
     fileConf <- either throwIO pure . parseRepresentation =<< BS.readFile fp
-    pure $ mergeWithDefault fileConf defaultConf
-
-mergeWithDefault ::  TraceConfig -> TraceConfig -> TraceConfig
-mergeWithDefault fileConf defaultConf =
-  TraceConfig
-    (if (not . Map.null) (tcOptions fileConf)
-      then tcOptions fileConf
-      else tcOptions defaultConf)
-    (tcForwarder fileConf)
-    (if isJust (tcNodeName fileConf)
-        then tcNodeName fileConf
-        else tcNodeName defaultConf)
-    (if isJust (tcPeerFrequency fileConf)
-        then tcPeerFrequency fileConf
-        else tcPeerFrequency defaultConf)
-    (if isJust (tcResourceFrequency fileConf)
-        then tcResourceFrequency fileConf
-        else tcResourceFrequency defaultConf)
-
+    pure $ mergeWithDefault fileConf
+  where
+    mergeWithDefault ::  TraceConfig -> TraceConfig
+    mergeWithDefault fileConf =
+      TraceConfig
+        (if (not . Map.null) (tcOptions fileConf)
+          then tcOptions fileConf
+          else tcOptions defaultConf)
+        (if isJust (tcForwarder fileConf)
+            then tcForwarder fileConf
+            else tcForwarder defaultConf)
+        (if isJust (tcNodeName fileConf)
+            then tcNodeName fileConf
+            else tcNodeName defaultConf)
+        (if isJust (tcPeerFrequency fileConf)
+            then tcPeerFrequency fileConf
+            else tcPeerFrequency defaultConf)
+        (if isJust (tcResourceFrequency fileConf)
+            then tcResourceFrequency fileConf
+            else tcResourceFrequency defaultConf)
+
```

Comment:
Is it an option to use `((<|>))` here? It would enhance readability, replacing conditional + `isJust` fields like so:
```
(tcForwarder fileConf <|> tcForwarder defaultConf)
```

Reply:
Is it possible that mappend from the Monoid instance for Maps could do similar for the first of these, or is there a possibility of absent configuration options that would be invalid to override with the defaults?

Reply:
Using `(<>)` extends to a set union of keys, which changes semantics, as you pointed out.

Reply:
Clever. I changed it